### PR TITLE
RavenDB-19021 : Sharding - Restore - Handle Compare Exchange and Subscriptions

### PR DESCRIPTION
**Pull request by aviv86**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-19021/Sharding-Restore-Handle-Compare-Exchange-and-Subscriptions

### Additional description

- introduce some changes and refactoring to `AbstractRestoreBackupTask` and `SingleShardRestoreBackupTask`
- set `ClusterTransactionIdBase64` and `Sharding.DatabaseId`
- add a new ctor to `ShardedDatabaseSmuggler` that will operate on `ISmugglerDestination` (instead of operating on `MultiShardedDestination` only) which `SingleShardRestoreBackupTask` will use now
- fix compare exchange tombstones handling in `ShardedDatabaseSmuggler` : use `CompareExchangeTombstones` actions (instead of `CompareExchange` actions), and remove compare exchange items directly if they're not atomic guards (instead of always writing them to stream when operating on `MultiShardedDestination`). Still missing atomic guard tombstones handling on sharded import/restore
- subscriptions will be handled by a single shard only
- tests adjustments

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreSnapshotTask.cs [hunk 1]
```diff
@@ -92,9 +92,9 @@ protected override async Task InitializeAsync()
             FilesToRestore.RemoveAt(0);
         }
 
-        protected override void OnAfterRestore()
+        protected override async Task OnAfterRestoreAsync()
         {
-            base.OnAfterRestore();
+            await base.OnAfterRestoreAsync();
```
Comment (karmeli87) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreSnapshotTask.cs:97]: Is there common code of the impl of `AbstractRestoreBackupTask` at `OnAfterRestoreAsync`
If no, it is better to make `OnAfterRestoreAsync` abstract
(I don't see any other call to `base.OnAfterRestoreAsync`)

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs [hunk 2]
```diff
@@ -64,21 +65,22 @@ protected override async Task RestoreAsync()
...
             using (serverContext.OpenReadTransaction())
             {
-                var databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, DatabaseName);
+                databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, DatabaseName);
                 databaseRecord.DatabaseState = DatabaseStateStatus.Normal;
-
-                RestoreSettings.DatabaseRecord = databaseRecord;
             }
+
+            await SaveDatabaseRecordAsync(DatabaseName, databaseRecord, null, Result, Progress);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs]: what is `null` here?

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs [hunk 3]
```diff
@@ -109,6 +112,7 @@ private void InitializeShardingConfiguration()
                     {
                         shardRestoreSetting.NodeTag
                     },
+                    ClusterTransactionIdBase64 = clusterTransactionIdBase64,
```
Comment (karmeli87) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs:115]: what about the orchestrator topology? let's make sure it is also identical (probably not only a restore thing)

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs [hunk 4]
```diff
@@ -27,42 +26,85 @@ internal class SingleShardRestoreBackupTask : RestoreBackupTask
...
+            return Task.CompletedTask;
         }
 
         protected override DatabaseRecord GetDatabaseRecord()
         {
             using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
             using (serverContext.OpenReadTransaction())
             {
-                var databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, _nonShardedDbName);
+                var databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, ShardHelper.ToDatabaseName(DatabaseName));
```
Comment (karmeli87) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs:55]: why do we need to override `GetDatabaseRecord` ? can we have the same method and simply pass the database name?

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs [hunk 5]
```diff
@@ -27,42 +26,85 @@ internal class SingleShardRestoreBackupTask : RestoreBackupTask
...
 
-        protected override void OnAfterRestore()
+        protected override async Task ImportLastBackupFileAsync(DocumentDatabase database, DatabaseDestination destination, JsonOperationContext context,
+            DatabaseSmugglerOptionsServerSide options, DatabaseRecord databaseRecord, string lastFilePath)
         {
-            SmugglerBase.EnsureProcessed(Result, skipped: false);
-            Progress.Invoke(Result.Progress);
-        }
+            if (_shardNumber > 0)
+                options.OperateOnTypes &= ~DatabaseItemType.Subscriptions;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs:73]: Add a sharding todo here, and create an issue to handle (or at least to discuss) having the most up-to-date database record

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 6]
```diff
@@ -802,14 +800,7 @@ public async ValueTask DisposeAsync()
...
                 //fire and forget, do not hold-up smuggler operations waiting for Raft command
-                DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Aviv, DevelopmentHelper.Severity.Normal, "try to find a better way to handle restoring identities on sharded database");
-                var databaseName = _database.Name;
-                if (_database is ShardedDocumentDatabase shardedDocumentDatabase)
-                {
-                    databaseName = shardedDocumentDatabase.ShardedDatabaseName;
-                }
-
-                await _database.ServerStore.SendToLeaderAsync(new UpdateClusterIdentityCommand(databaseName, _identities, false, RaftIdGenerator.NewId()));
+                await _database.ServerStore.SendToLeaderAsync(new UpdateClusterIdentityCommand(_database.Name, _identities, false, RaftIdGenerator.NewId()));
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: what is `false` here?

Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: BTW, the comment above isn't true here... we are awaiting.

File: src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs [hunk 7]
```diff
@@ -248,16 +263,43 @@ await foreach (var kvp in _source.GetCompareExchangeValuesAsync())
...
+                    {
+                        result.CompareExchangeTombstones.ErroredCount++;
+                        continue;
+                    }
+
+                    try
+                    {
+                        if (ClusterTransactionCommand.IsAtomicGuardKey(kvp.Key.Key, out _))
+                        {
+                            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Aviv, DevelopmentHelper.Severity.Normal, "handle atomic guard tombstones import");
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs]: Do we have a test to check this scenario? if not, let's create an issue and move on

File: src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs [hunk 8]
```diff
@@ -40,7 +41,21 @@ public class ShardedDatabaseSmuggler : SmugglerBase
             long operationId,
             Action<IOperationProgress> onProgress = null, 
             CancellationToken token = default) : 
-            base(source, new MultiShardedDestination(source, databaseContext, handler, operationId), server.Server.Time, jsonOperationContext, options, result, onProgress, token)
+            this(source, new MultiShardedDestination(source, databaseContext, handler, operationId), jsonOperationContext, databaseRecord, server, options, result, onProgress, token)
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs]: let's have a single ctor, and pass different destination from the caller

Comment (aviv86) [reply src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs]: done

File: src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs (hunk 7 above)
Comment (aviv86) [reply src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs]: https://issues.hibernatingrhinos.com/issue/RavenDB-19201/Sharding-Import-handle-atomic-guard-tombstones

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 6 above)
Comment (aviv86) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: `false` here is `force` update
removed comment

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs (hunk 5 above)
Comment (aviv86) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs:73]: https://issues.hibernatingrhinos.com/issue/RavenDB-19202/Sharding-Backup-Restore-consider-using-the-most-up-to-date-database-record

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs (hunk 4 above)
Comment (aviv86) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs:55]: common implementation of `GetDatabaseRecord` is creating a new instance

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs (hunk 3 above)
Comment (aviv86) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs:115]: done

File: src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/ShardedRestoreOrchestrationTask.cs]: null is `databaseValues`
added name identifier

File: src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreSnapshotTask.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreSnapshotTask.cs:97]: yes, there's common code of `OnAfterRestoreAsync` at `AbstractRestoreBackupTask`, used by non-sharded backup restore and by snapshot restore

**Review by ppekrol — APPROVED** [review]

**Pull request merged**